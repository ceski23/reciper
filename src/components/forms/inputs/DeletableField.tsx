import styled from '@emotion/styled';
import { ComponentPropsWithoutRef, forwardRef } from 'react';

import { ReactComponent as DeleteIcon } from 'assets/common/delete.svg';

import { Input } from 'components/common/Input';

import { color } from 'utils/styles/theme';

interface NumberFieldProps {
  onDeleteClick: () => void;
}

const StyledInput = styled(Input)`
  border-right: none;
  border-top-right-radius: 0;
  border-bottom-right-radius: 0;
`;

const DeleteButton = styled.button`
  border-radius: 0 10px 10px 0;
  border: 1px solid ${color('textalt')};
  padding: 8px 20px;
  display: flex;
  align-items: center;
  justify-content: center;
  background-color: ${color('backgroundAltHover')};
  color: ${color('textalt')};
  cursor: pointer;

  &:hover {
    background-color: ${color('backgroundAlt')};
  }

  svg {
    width: 15px;
    height: 15px;
  }
`;

const InputContainer = styled.div`
  display: flex;
  flex-direction: row-reverse;
`;

export const DeletableField = forwardRef<HTMLInputElement, NumberFieldProps & ComponentPropsWithoutRef<'input'>>(({
  onDeleteClick, ...props
}, ref) => (
  <InputContainer>
    <DeleteButton type="button" onClick={onDeleteClick}>
      <DeleteIcon />
    </DeleteButton>

    <StyledInput ref={ref} {...props} />
  </InputContainer>
));
