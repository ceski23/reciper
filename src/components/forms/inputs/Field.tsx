import styled from '@emotion/styled';
import React, { forwardRef } from 'react';

import { color } from 'utils/styles/theme';

const DeleteButton = styled.button`
  border-radius: 0 10px 10px 0;
  border: 1px solid ${color('textalt')};
  padding: 8px 10px;
  display: flex;
  align-items: center;
  justify-content: center;
  background-color: ${color('backgroundAltHover')};
  color: ${color('textalt')};
  cursor: pointer;

  &:hover {
    background-color: ${color('backgroundAlt')};
  }

  svg {
    width: 15px;
    height: 15px;
  }
`;

const StyledField = styled.input`
  font-family: inherit;
  font-size: 14px;
  font-weight: 500;
  padding: 10px 15px;
  border-radius: 10px;
  border: 1px solid ${color('textalt')};
  background-color: ${color('backgroundInput')};
  color: ${color('textalt')};
  flex: 1;

  &::placeholder {
    color: ${color('textalt')};
    opacity: 1;
  }

  ${DeleteButton} + & {
    border-radius: 10px 0 0 10px;
    border-right: none;
  }
`;

const FieldContainer = styled.div`
  display: flex;
  flex-direction: column;
  width: 100%;
`;

const LabelText = styled.label`
  margin-bottom: 10px;
  font-weight: 700;
  font-size: 16px;
`;

const Error = styled.p`
  font-size: 13px;
  color: red;
`;

const InputContainer = styled.div`
  display: flex;
  flex-direction: row-reverse;
`;

const Required = styled.i`
  color: #ff0000;
`;

export type InputProps =
  React.DetailedHTMLProps<React.InputHTMLAttributes<HTMLInputElement>, HTMLInputElement>;

type FieldProps = Pick<InputProps, 'id' | 'aria-invalid' | 'aria-required' | 'aria-describedby'>;

interface Props {
  label?: string;
  error?: string;
  deleteIcon?: React.FunctionComponent<React.SVGProps<SVGSVGElement>>;
  onDeleteClick?: () => void;
  required?: boolean;
  id: string;
  render?: (fieldProps: FieldProps) => React.ReactNode;
}

export const Field = forwardRef<HTMLInputElement, InputProps & Props>(({
  error,
  label,
  children,
  deleteIcon: Icon,
  onDeleteClick,
  required = false,
  id,
  render,
  ...props
}, ref) => (
  <FieldContainer>
    {label && (
      <LabelText htmlFor={id}>
        {label}
        {required && <Required aria-hidden>*</Required>}
      </LabelText>
    )}

    {render ? render({
      id,
      'aria-invalid': !!error,
      'aria-required': required,
      'aria-describedby': error ? `${id}-error` : undefined,
    }) : (
      <InputContainer>
        {Icon && onDeleteClick && (
          <DeleteButton type="button" onClick={onDeleteClick}>
            <Icon />
          </DeleteButton>
        )}

        <StyledField
          id={id}
          aria-invalid={!!error}
          aria-required={required}
          aria-describedby={error ? `${id}-error` : undefined}
          ref={ref}
          {...props}
        />
      </InputContainer>
    )}

    {error && <Error role="alert" id={`${id}-error`}>{error}</Error>}
  </FieldContainer>
));
