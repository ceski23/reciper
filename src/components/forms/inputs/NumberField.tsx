import styled from '@emotion/styled';
import { ComponentPropsWithoutRef, forwardRef } from 'react';

import { Input } from 'components/common/Input';

import { color } from 'utils/styles/theme';

interface NumberFieldProps {
  step?: number;
  onChange: (value: number) => void;
  value?: number;
}

const StepperContainer = styled.div`
  display: flex;
  flex-direction: row;
  width: 100%;
`;

const StyledInput = styled(Input)`
  flex: 1;
  border-left: none;
  border-right: none;
  border-radius: 0;
  text-align: center;
`;

const InputButton = styled.button`
  border: 1px solid ${color('textalt')};
  padding: 8px 20px;
  display: flex;
  align-items: center;
  justify-content: center;
  background-color: ${color('backgroundAltHover')};
  color: ${color('textalt')};
  cursor: pointer;
  font-weight: 700;
  font-size: 20px;

  &:hover {
    background-color: ${color('backgroundAlt')};
  }

  svg {
    width: 15px;
    height: 15px;
  }
`;

type InputProps = Omit<ComponentPropsWithoutRef<'input'>, keyof NumberFieldProps>;

const normalizeNumber = (value: string, fallback?: number): number => {
  const newNumber = Number(value);
  const newValue = Number.isNaN(newNumber) ? (fallback ?? 0) : newNumber;

  return newValue;
};

export const NumberField = forwardRef<HTMLInputElement, NumberFieldProps & InputProps>(({
  onChange, value = 0, step = 1, ...props
}, ref) => {
  const handleChange = (rawValue: string) => {
    const newNumber = normalizeNumber(rawValue, Number(value));
    onChange(newNumber);
  };

  return (
    <StepperContainer>
      <InputButton
        type="button"
        onClick={() => handleChange(String(value - step))}
        style={{ borderRadius: '10px 0 0 10px' }}
      >-
      </InputButton>

      <StyledInput
        ref={ref}
        value={value}
        onChange={(e) => handleChange(e.currentTarget.value)}
        type="text"
        inputMode="numeric"
        {...props}
      />

      <InputButton
        type="button"
        onClick={() => handleChange(String(value + step))}
        style={{ borderRadius: '0 10px 10px 0' }}
      >+
      </InputButton>
    </StepperContainer>
  );
});
