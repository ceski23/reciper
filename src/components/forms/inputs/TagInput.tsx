import styled from '@emotion/styled';
import {
  ChangeEventHandler, KeyboardEventHandler, useState, VFC,
} from 'react';

import { ReactComponent as DeleteIcon } from 'assets/common/delete.svg';

import { color } from 'utils/styles/theme';

interface Props {
  onChange: (value: string[]) => void;
  onBlur: () => void;
  value: string[];
  name: string;
  id?: string;
}

const StyledField = styled.input`
  flex: 1;
  font-family: inherit;
  font-size: 14px;
  font-weight: 500;
  border: none;
  padding: 5px 10px;
  min-width: 0;
  background-color: ${color('backgroundInput')};

  &::placeholder {
    color: ${color('textalt')};
    opacity: 1;
  }

  &:focus {
    outline: none;
  }
`;

const Tag = styled.span`
  display: flex;
  gap: 10px;
  justify-content: center;
  align-items: center;
  background-color: ${color('backgroundAltHover')};
  padding: 5px 10px;
  border-radius: 10px;
  font-size: 14px;
  font-weight: 500;
`;

const TagDeleteButton = styled.button`
  border-radius: 50%;
  border: none;
  display: flex;
  width: 20px;
  height: 20px;
  justify-content: center;
  align-items: center;
  font-weight: 500;
  padding: 5px;
  background-color: ${color('backgroundAlt')};
  color: ${color('text')};
`;

const TagsContainer = styled.div`
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
  gap: 10px;
  padding: 10px 15px;
  border-radius: 10px;
  border: 1px solid ${color('textalt')};
  background-color: ${color('backgroundInput')};
`;

export const TagInput: VFC<Props> = ({
  onBlur, onChange, name, value, id, ...props
}) => {
  const [internalValue, setInternalValue] = useState<string>('');

  const handleKeyDown: KeyboardEventHandler<HTMLInputElement> = (event) => {
    const data = internalValue.trim();

    if (data.length > 0 && [',', 'Enter'].includes(event.key) && !value.includes(data)) {
      event.preventDefault();
      onChange([...value, data]);
      setInternalValue('');
    }

    if (data.length === 0 && event.key === 'Backspace') {
      event.preventDefault();
      onChange(value.slice(0, -1));
    }
  };

  const handleChange: ChangeEventHandler<HTMLInputElement> = (event) => {
    setInternalValue(event.currentTarget.value);
  };

  const deleteTag = (tag: string) => {
    onChange(value.filter((t) => t !== tag));
  };

  return (
    <TagsContainer>
      {value.map((tag) => (
        <Tag key={tag}>
          {tag}
          <TagDeleteButton aria-label={tag} type="button" onClick={() => deleteTag(tag)}>
            <DeleteIcon />
          </TagDeleteButton>
        </Tag>
      ))}

      <StyledField
        {...props}
        onChange={handleChange}
        onKeyDown={handleKeyDown}
        value={internalValue}
        name={name}
        onBlur={onBlur}
        placeholder="Podaj tag..."
        enterKeyHint="next"
        id={id}
      />
    </TagsContainer>
  );
};
