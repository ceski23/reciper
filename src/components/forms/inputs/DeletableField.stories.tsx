import { ComponentMeta, ComponentStory } from '@storybook/react';
import { ChangeEvent, useState } from 'react';

import { DeletableField } from 'components/forms/inputs/DeletableField';

export default {
  title: 'Components/Form/DeletableField',
  component: DeletableField,
} as ComponentMeta<typeof DeletableField>;

const Template: ComponentStory<typeof DeletableField> = ({ value, ...args }) => {
  const [inputValue, setInputValue] = useState(value);

  const handleChange = (event: ChangeEvent<HTMLInputElement>) => {
    const newValue = event.currentTarget.value;
    setInputValue(newValue);
    args.onChange?.(event);
  };

  return <DeletableField {...args} value={inputValue} onChange={handleChange} />;
};

export const Default = Template.bind({});
Default.args = {};
