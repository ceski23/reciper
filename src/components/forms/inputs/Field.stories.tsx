import { ComponentMeta, ComponentStory } from '@storybook/react';

import { Field } from 'components/forms/inputs/Field';

export default {
  title: 'Components/Form/Field',
  component: Field,
} as ComponentMeta<typeof Field>;

const Template: ComponentStory<typeof Field> = (args) => <Field {...args} />;

export const Default = Template.bind({});
Default.args = {};

export const WithLabel = Template.bind({});
WithLabel.args = {
  label: 'Field label',
  required: true,
};

export const WithError = Template.bind({});
WithError.args = {
  label: 'Field label',
  required: true,
  error: 'Some error message',
};
