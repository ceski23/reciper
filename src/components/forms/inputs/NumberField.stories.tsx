import { ComponentMeta, ComponentStory } from '@storybook/react';
import { useState } from 'react';

import { NumberField } from 'components/forms/inputs/NumberField';

export default {
  title: 'Components/Form/NumberField',
  component: NumberField,
} as ComponentMeta<typeof NumberField>;

const Template: ComponentStory<typeof NumberField> = ({ value, ...args }) => {
  const [inputValue, setInputValue] = useState(value);

  const handleChange = (newValue: number) => {
    setInputValue(newValue);
    args.onChange(newValue);
  };

  return <NumberField {...args} value={inputValue} onChange={handleChange} />;
};

export const Default = Template.bind({});
Default.args = {};

export const WithStep = Template.bind({});
WithStep.args = {
  step: 5,
};
