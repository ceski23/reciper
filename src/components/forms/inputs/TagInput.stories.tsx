import { ComponentMeta, ComponentStory } from '@storybook/react';
import { useState } from 'react';

import { TagInput } from 'components/forms/inputs/TagInput';

export default {
  title: 'Components/Form/TagInput',
  component: TagInput,
} as ComponentMeta<typeof TagInput>;

const Template: ComponentStory<typeof TagInput> = ({ value, ...args }) => {
  const [inputValue, setInputValue] = useState(value);

  const handleChange = (newValue: string[]) => {
    setInputValue(newValue);
    args.onChange(newValue);
  };

  return <TagInput {...args} value={inputValue} onChange={handleChange} />;
};

export const Default = Template.bind({});
Default.args = {
  name: 'test',
  value: [],
};
