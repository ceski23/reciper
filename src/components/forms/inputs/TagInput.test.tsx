import { screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { renderWithProviders } from 'test/utils';

import { TagInput } from 'components/forms/inputs/TagInput';

describe('TagInput', () => {
  it('should render correctly', () => {
    const { asFragment } = renderWithProviders(
      <TagInput name="test" value={['tag1', 'tag2']} onBlur={vi.fn()} onChange={vi.fn()} />,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('should remove last tag', async () => {
    const handleChange = vi.fn();
    const user = userEvent.setup();

    renderWithProviders(
      <TagInput
        name="test"
        value={['tag1', 'tag2']}
        onBlur={vi.fn()}
        onChange={handleChange}
      />,
    );

    // Remove using backspace
    const input = screen.getByRole('textbox');
    await user.type(input, '{backspace}');
    expect(handleChange).toBeCalledWith(['tag1']);

    // Remove using X button
    const tag1Button = screen.getByRole('button', { name: /tag1/i });
    await user.click(tag1Button);
    expect(handleChange).toBeCalledWith(['tag2']);
  });

  it('should add new tag', async () => {
    const handleChange = vi.fn();
    const user = userEvent.setup();

    renderWithProviders(
      <TagInput
        name="test"
        value={['tag1']}
        onBlur={vi.fn()}
        onChange={handleChange}
      />,
    );

    const input = screen.getByRole('textbox');

    // Accept tag using enter
    await user.type(input, 'tag2{enter}');
    expect(handleChange).toBeCalledWith(['tag1', 'tag2']);

    // Accept tag using comma
    await user.type(input, 'tag3,');
    expect(handleChange).toBeCalledWith(['tag1', 'tag3']);
  });
});
