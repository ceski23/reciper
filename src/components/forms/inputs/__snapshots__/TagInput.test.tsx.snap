// Vitest Snapshot v1

exports[`TagInput > should render correctly 1`] = `
<DocumentFragment>
  .emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-flex-wrap: wrap;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  gap: 10px;
  padding: 10px 15px;
  border-radius: 10px;
  border: 1px solid #365463;
  background-color: #fff;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 10px;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: #cfdce2;
  padding: 5px 10px;
  border-radius: 10px;
  font-size: 14px;
  font-weight: 500;
}

.emotion-4 {
  border-radius: 50%;
  border: none;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 20px;
  height: 20px;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-weight: 500;
  padding: 5px;
  background-color: #dfe8ec;
  color: #023;
}

.emotion-10 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  font-family: inherit;
  font-size: 14px;
  font-weight: 500;
  border: none;
  padding: 5px 10px;
  min-width: 0;
  background-color: #fff;
}

.emotion-10::-webkit-input-placeholder {
  color: #365463;
  opacity: 1;
}

.emotion-10::-moz-placeholder {
  color: #365463;
  opacity: 1;
}

.emotion-10:-ms-input-placeholder {
  color: #365463;
  opacity: 1;
}

.emotion-10::placeholder {
  color: #365463;
  opacity: 1;
}

.emotion-10:focus {
  outline: none;
}

<div
    class="emotion-0 emotion-1"
  >
    <span
      class="emotion-2 emotion-3"
    >
      tag1
      <button
        aria-label="tag1"
        class="emotion-4 emotion-5"
        type="button"
      >
        <svg
          viewBox="0 0 512.001 512.001"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M294.111 256.001L504.109 46.003c10.523-10.524 10.523-27.586 0-38.109-10.524-10.524-27.587-10.524-38.11 0L256 217.892 46.002 7.894C35.478-2.63 18.416-2.63 7.893 7.894s-10.524 27.586 0 38.109l209.998 209.998L7.893 465.999c-10.524 10.524-10.524 27.586 0 38.109 10.524 10.524 27.586 10.523 38.109 0L256 294.11l209.997 209.998c10.524 10.524 27.587 10.523 38.11 0 10.523-10.524 10.523-27.586 0-38.109L294.111 256.001z"
            fill="currentColor"
          />
        </svg>
      </button>
    </span>
    <span
      class="emotion-2 emotion-3"
    >
      tag2
      <button
        aria-label="tag2"
        class="emotion-4 emotion-5"
        type="button"
      >
        <svg
          viewBox="0 0 512.001 512.001"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M294.111 256.001L504.109 46.003c10.523-10.524 10.523-27.586 0-38.109-10.524-10.524-27.587-10.524-38.11 0L256 217.892 46.002 7.894C35.478-2.63 18.416-2.63 7.893 7.894s-10.524 27.586 0 38.109l209.998 209.998L7.893 465.999c-10.524 10.524-10.524 27.586 0 38.109 10.524 10.524 27.586 10.523 38.109 0L256 294.11l209.997 209.998c10.524 10.524 27.587 10.523 38.11 0 10.523-10.524 10.523-27.586 0-38.109L294.111 256.001z"
            fill="currentColor"
          />
        </svg>
      </button>
    </span>
    <input
      class="emotion-10 emotion-11"
      enterkeyhint="next"
      name="test"
      placeholder="Podaj tag..."
      value=""
    />
  </div>
</DocumentFragment>
`;
